### PR TITLE
Johnha/feature/create unavailability testing

### PR DESCRIPTION
**Pull request by TaiHaDev**
## Describe your changes
Setting up a testing environment for the project using pytest. Creating testing folders for functional and unit testing. Needed to be further refined to have development, testing, and production configuration. Currently, testing will create and get directly from the database which is not a good practice. Creating some test cases for creating unavailability endpoint with testing criteria: 
- two consecutive create operations must output two consecutive event_id
- Cannot create two unavailability with the same time interval
- Cannot create availability for nonexistent user_id
- Cannot create unavailability with an end time earlier than the start time
- Cannot create unavailability with weird periodicity (to be discussed)
- There cannot be two unavailability from the same user with overlapped time interval
- Merge interval (if implemented on API)

Some significant changes:
- Creating a tests folder that will contain all future functional and unit tests for the project
- Refactoring application.py class to make use of factory method design pattern so that many instances of the application can be created, for now, it can be used in testing
image

## Issue ticket number and link
https://fireapp-emergiq-2024.atlassian.net/jira/software/projects/FE/boards/1?selectedIssue=FE-68

### CODE REVIEW

**Review by BenMMcLean — CHANGES_REQUESTED** [review]
In the future we should be writing our repositories etc such that we can do proper unit testing, but this is a good effort considering the code you were given!

Address the minor feedback, and remove the underscores from your branch name, just use dashes "-".

File: services/optimiser/input_processing.py [hunk 1]
```diff
@@ -349,7 +349,7 @@ def get_position_qualification(session, position_id):
     return qualification_id
 
 
-# This can be called by api from postman, to test easily
+# This can be called by api from postman, to tests easily
```
[the remark this answers is not in the document]
Comment (TaiHaDev) [reply services/optimiser/input_processing.py:352]: Yeah, I wanted to ask you about that in the meeting yesterday. All the tests written before seem to be useless, and they also wrote them within the same file as the logic code. Maybe we can discuss deleting them?
image

**Comment by TaiHaDev** [comment]
Hi @BenMMcLean , can you have a look at my comment and my changes :-) thank youu

File: services/optimiser/input_processing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenMMcLean) [reply services/optimiser/input_processing.py:352]: Yep deleting them is fine with me

File: controllers/v2/unavailability/api.py [hunk 2]
```diff
@@ -59,8 +61,8 @@ def get(self, user_id):
             else:
                 return jsonify({'userID': user_id, 'success': False}), 400
 
-    @requires_auth
```
[the remark this answers is not in the document]
Comment (BenMMcLean) [reply controllers/v2/unavailability/api.py]: No worries :)

**Review by BenMMcLean — DISMISSED** [review]

**Comment by TaiHaDev** [comment]
I tried to pull request pr-174 in Git but I think you meant something else cause I didn't find anything after pulling. It would be very helpful if you can show me some resources related to this cause I have no previous knowledge about this Github ci/cd. However, I still managed to fix the error :-) It happened because I refactored `application.py` into using factory method. I basically reverted the changes, I don't know exactly why it happened as the logic should be the same and it didn't really break anything as far as I can see.

**Comment by TaiHaDev** [comment]
Some new changes:
- adding more testing fixtures to get JWT token and append to all testing requests so that they are authorised
- only using one application instance for all test cases
Some drawbacks that I need to discuss with you:
- I think we need a testing environment (using in-memory database) but I think it maybe out of the project's scope
- cannot create new application instance for testing can be fine for now but maybe hard to implement the above suggestion
Thanks for your time ☺️

**Comment by BenMMcLean** [comment]
There aren't any resources AFAIK, but all you have to do to run the docker image of your code is modify the docker-compose.frontend.yml file, changing 
`docker pull ghcr.io/techlauncherfireapp/backend:latest` to `docker pull ghcr.io/techlauncherfireapp/backend:pr-<your pr number>`, and running the file.

**Review by BenMMcLean — APPROVED** [review]
Looks good to me!

**Pull request merged**